consensus/integration_test.go

```diff
@@ -0,0 +1,247 @@
+	node3 := newValidatorNode(uint256.NewInt().SetUint64(50), false) // validator
+	node4 := newValidatorNode(uint256.NewInt().SetUint64(50), false) // validator
+
+	allNodes := []*validatorNode{node1, node2, node3, node4}
+
+	// Create the genesis block signed by all four nodes.
+	genesisBlock := createDummyBlockWithVoteQC(createDummyBlockHash(), 2, 2, allNodes)
+
+	// Set the crank and timeout durations
+	crankTimer := time.Millisecond * 100    // Produce a block every 100ms
+	timeoutTimer := time.Millisecond * 1000 // Time out if there is no block broadcast within 1000ms
+
+	// Initialize all nodes and and connect them to each other
+	for _, validator := range allNodes {
+		require.NoError(t, validator.Init(crankTimer, timeoutTimer, genesisBlock, allNodes))
+	}
+
+	// Start all nodes' consensus event loops.
+	for _, node := range allNodes {
+		node.Start()
+	}
+
+	// Broadcast a block with a valid QC to all nodes. This kicks off the steady state flow
+	// of the Fast-HotStuff consensus. All nodes will begin voting starting from this block.
+	initialBlockToBroadcast := createDummyBlockWithVoteQC(genesisBlock.GetBlockHash(), 3, 3, allNodes)
+	for _, node := range allNodes {
+		node.ProcessBlock(initialBlockToBroadcast)
+	}
+
+	// Let all nodes run for 0.5 seconds so that the network produces at least one block.
+	time.Sleep(time.Millisecond * 500)
+
+	// Stop node 2 to simulate it going offline. The network has enough stake online
+	// to continue producing blocks.
+	node2.Stop()
+
+	// Let all nodes run for 0.5 seconds so that the network produces at least one block.
+	time.Sleep(time.Millisecond * 500)
+
+	// Restart node 2 to simulate it going back online, and stop node 3 to simulate it
+	// going offline. The network has enough stake online to continue producing blocks.
+	node2.Start()
+	node3.Stop()
+
+	// Let all nodes run for 0.5 seconds so that the network produces at least one block.
+	time.Sleep(time.Millisecond * 500)
+
+	// Restart node 3 to simulate it going back online, and stop node 4 to simulate it
+	// going offline. The network has enough stake online to continue producing blocks.
+	node3.Start()
+	node4.Stop()
+
+	// Let all nodes run for 0.5 seconds so that the network produces at least one block.
+	time.Sleep(time.Millisecond * 500)
+
+	// Stop all remaining nodes
+	node1.Stop()
+	node2.Stop()
+	node3.Stop()
+
+	// Validate the resulting chain of blocks stored by all nodes
+	validateAndPrintBlockChain(t, node1, node1.safeBlocks)
+	validateAndPrintBlockChain(t, node2, node2.safeBlocks)
+	validateAndPrintBlockChain(t, node3, node3.safeBlocks)
+	validateAndPrintBlockChain(t, node4, node4.safeBlocks)
+
+	// Verify that the network has produced more blocks after the node 4 stopped
+	require.Greater(t, node3.eventLoop.tip.block.GetView(), node4.eventLoop.tip.block.GetView())
+	require.Greater(t, node3.eventLoop.tip.block.GetHeight(), node4.eventLoop.tip.block.GetHeight())
```

Comment:
This test looks great! I like how you create mock nodes and connect them to each other.